Update with all core features (#24037)
backstage/backstage: docs/backend-system/building-backends/08-migrating.md

```

> You can track the progress of the module migration efforts [here](https://github.com/backstage/backstage/issues/19476).

### The Search Plugin

A basic installation of the Search plugin will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-search-backend/alpha'));
/* highlight-add-end */
```

> Note: this will use the Lunr search engine which stores its index in memory

#### Search Engines

> The following sections outline how you can add other Search engines

Reviewer: ```suggestion
The following sections outline how you can add other Search engines than the default lunr engine.
```

##### Postgres

An installation of the Search plugin using the Postgres search engine will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-search-backend/alpha'));
backend.add(import('@backstage/plugin-search-backend-module-pg/alpha'));
/* highlight-add-end */
```

##### Elasticsearch

A basic installation of the Search plugin using the Elasticsearch search engine will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-search-backend/alpha'));
backend.add(
  import('@backstage/plugin-search-backend-module-elasticsearch/alpha'),
);
/* highlight-add-end */
```

#### Search Collators

> The following sections outline how you add the Catalog and TechDocs search collators

Reviewer: ```suggestion
The following sections outline how you add search collators (input sources for the search indexing process).
```

just to make it easier to expand in the future without missing updating the text here too

##### Catalog

A basic installation of the Search plugin with the Catalog collator will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-search-backend/alpha'));
backend.add(import('@backstage/plugin-search-backend-module-catalog/alpha'));
/* highlight-add-end */
```

##### TechDocs

A basic installation of the Search plugin with the TechDocs collator will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-search-backend/alpha'));
backend.add(import('@backstage/plugin-search-backend-module-techdocs/alpha'));
/* highlight-add-end */
```

> ### The Permissions Plugin

Reviewer: ```suggestion
### The Permission Plugin
```

i keep forgetting whether this is pluralized or not 😆

A basic installation of the Permission plugin will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-permission-backend/alpha'));
backend.add(
  import('@backstage/plugin-permission-backend-module-allow-all-policy'),
);
/* highlight-add-end */
```

> > Note: the above includes a default allow all policy, if that is not what you want do not add the 2nd line.

Reviewer: ```suggestion
> Note: The above example includes a default allow-all policy. If that is not what you want, do not add the second line and instead investigate one of the options below.
```

#### Custom Permission Policy

In order to add your own permission policy you'll need to do the following:

```ts
import { createBackendModule } from '@backstage/backend-plugin-api';
import { BackstageIdentityResponse } from '@backstage/plugin-auth-node';
import {
  PolicyDecision,
  AuthorizeResult,
} from '@backstage/plugin-permission-common';
import {
  PermissionPolicy,
  PolicyQuery,
} from '@backstage/plugin-permission-node';
import { policyExtensionPoint } from '@backstage/plugin-permission-node/alpha';

> class ExampleAllowAllPermissionPolicy implements PermissionPolicy {

Reviewer: It's probably unfortunate to explicitly name this `AllowAll`. Imagine that this will be copy pasted a lot and then only the contents of the `handle` method updated. Maybe name the class `CustomPermissionPolicy` to begin with?

Author: Yeah, that makes sense as I do `policy.setPolicy(new CustomPermissionPolicy());` so this doesn't even match what I have across the example.

Author: Fixed up 👍

  async handle(
    _request: PolicyQuery,
    _user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    return {

>       result: AuthorizeResult.ALLOW,

Reviewer: and then put a comment in here instead of actual code, something like `// TODO: Add code here that inspects the incoming request and user, and returns AuthorizeResult.ALLOW, AuthorizeResult.DENY, or AuthorizeResult.CONDITIONAL as needed. See the docs at <url> for more information`

Reviewer: at that point perhaps the _request and _user can drop their underscores too - it's one of those tricky ones where it's not certain which is more helpful to end users who aren't familiar with ts coding :)

    };
  }
}

const customPermissionBackendModule = createBackendModule({
  pluginId: 'permission',

>   moduleId: 'allow-all-policy',

Reviewer: ```suggestion
  moduleId: 'custom-policy',
```

  register(reg) {
    reg.registerInit({
      deps: { policy: policyExtensionPoint },
      async init({ policy }) {

>         policy.setPolicy(new ExampleAllowAllPermissionPolicy());

Reviewer: ```suggestion
        policy.setPolicy(new CustomPermissionPolicy());
```

      },
    });
  },
});

const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-permission-backend/alpha'));
backend.add(customPermissionBackendModule);
/* highlight-add-end */
```

### The TechDocs Plugin

A basic installation of the TechDocs plugin will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */

> backend.add(import('@backstage/plugin-search-backend-module-catalog/alpha'));

Reviewer 2: This is the catalog import url, but should be `@backstage/plugin-search-backend-module-techdocs/alpha`.

Author: Oh, nice catch! Too much copy-pasta!

Author: Fixed up 👍

  async handle(
    request: PolicyQuery,
    user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    // TODO: Add code here that inspects the incoming request and user, and returns AuthorizeResult.ALLOW, AuthorizeResult.DENY, or AuthorizeResult.CONDITIONAL as needed. See the docs at https://backstage.io/docs/permissions/writing-a-policy for more information

    return {
      result: AuthorizeResult.ALLOW,
    };
  }
}

const customPermissionBackendModule = createBackendModule({
  pluginId: 'permission',
  moduleId: 'custom-policy',
  register(reg) {
    reg.registerInit({
      deps: { policy: policyExtensionPoint },
      async init({ policy }) {
        policy.setPolicy(new CustomPermissionPolicy());
      },
    });
  },
});

const backend = createBackend();

// Other plugins...

/* highlight-add-start */
backend.add(import('@backstage/plugin-permission-backend/alpha'));
backend.add(customPermissionBackendModule);
/* highlight-add-end */
```

### The TechDocs Plugin

A basic installation of the TechDocs plugin will look as follows:

```ts title="packages/backend/src/index.ts"
const backend = createBackend();

// Other plugins...

/* highlight-add-start */

> backend.add(import('@backstage/plugin-search-backend-module-techdocs/alpha'));

Reviewer 2: I see you addressed my comment before I corrected my url recommendation. 😊 Sorry, that was the search techdocs url. The main techdocs one is: `@backstage/plugin-techdocs-backend/alpha`.

Author: You are absolutely correct, trying to move too fast! Really want this update merged as I know it will help. many Adopters but it needs to be right so really thankful you followed up!!!

Author: Also, this has been fixed 👍
